File: wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/rapidreactcommandbot/subsystems/Shooter.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.examples.rapidreactcommandbot.subsystems;

import static edu.wpi.first.wpilibj2.command.Commands.parallel;
import static edu.wpi.first.wpilibj2.command.Commands.waitUntil;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.examples.rapidreactcommandbot.Constants.ShooterConstants;
import edu.wpi.first.wpilibj.motorcontrol.PWMSparkMax;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.CommandScheduler;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

public class Shooter extends SubsystemBase implements AutoCloseable {
  private final PWMSparkMax m_shooterMotor = new PWMSparkMax(ShooterConstants.kShooterMotorPort);
  private final PWMSparkMax m_feederMotor = new PWMSparkMax(ShooterConstants.kFeederMotorPort);
  private final Encoder m_shooterEncoder =
      new Encoder(
          ShooterConstants.kEncoderPorts[0],
          ShooterConstants.kEncoderPorts[1],
          ShooterConstants.kEncoderReversed);
  private final SimpleMotorFeedforward m_shooterFeedforward =
      new SimpleMotorFeedforward(
          ShooterConstants.kSVolts, ShooterConstants.kVVoltSecondsPerRotation);
  private final PIDController m_shooterFeedback = new PIDController(ShooterConstants.kP, 0.0, 0.0);

  /** The shooter subsystem for the robot. */
  public Shooter() {
    m_shooterFeedback.setTolerance(ShooterConstants.kShooterToleranceRPS);
    m_shooterEncoder.setDistancePerPulse(ShooterConstants.kEncoderDistancePerPulse);

    // Set default command to turn off both the shooter and feeder motors, and then idle
    setDefaultCommand(
        runOnce(
                () -> {
                  m_shooterMotor.set(0);
                  m_feederMotor.set(0);
                })
            .andThen(run(() -> {}))
            .ignoringDisable(true)
            .withName("Idle"));
  }

  /**
   * Returns a command to shoot the balls currently stored in the robot. Spins the shooter flywheel
   * up to the specified setpoint, and then runs the feeder motor.
   *
   * @param setpointRotationsPerSecond The desired shooter velocity
   */
  public Command shootCommand(double setpointRotationsPerSecond) {
    return parallel(
            // Run the shooter flywheel at the desired setpoint using feedforward and feedback
            run(
                () ->
                    m_shooterMotor.setVoltage(
                        m_shooterFeedforward.calculate(setpointRotationsPerSecond)
                            + m_shooterFeedback.calculate(
                                m_shooterEncoder.getRate(), setpointRotationsPerSecond))),
            // Wait until the shooter has reached the setpoint, and then run the feeder
            waitUntil(m_shooterFeedback::atSetpoint)
                .andThen(() -> m_feederMotor.set(ShooterConstants.kFeederSpeed)))
        .withName("Shoot");
  }

  @Override
  public void close() {
    m_feederMotor.close();
    m_shooterEncoder.close();
    m_shooterFeedback.close();
    m_shooterMotor.close();
    CommandScheduler.getInstance().unregisterSubsystem(this);
  }

  public double getShooterVelocity() {
    return m_shooterEncoder.getRate();
  }
}

File: wpilibcExamples/src/main/cpp/examples/RapidReactCommandBot/cpp/subsystems/Shooter.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/Shooter.h"

#include <frc2/command/Commands.h>
#include <frc2/command/button/Trigger.h>

Shooter::Shooter() {
  m_shooterFeedback.SetTolerance(ShooterConstants::kShooterTolerance.value());
  m_shooterEncoder.SetDistancePerPulse(
      ShooterConstants::kEncoderDistancePerPulse);

  SetDefaultCommand(RunOnce([this] {
                      m_shooterMotor.Set(0.0);
                      m_feederMotor.Set(0.0);
                    })
                        .AndThen(Run([] {}))
                        .IgnoringDisable(true)
                        .WithName("Idle"));
}

units::turns_per_second_t Shooter::GetShooterVelocity() {
  return units::turns_per_second_t(m_shooterEncoder.GetRate());
}

frc2::CommandPtr Shooter::ShootCommand(units::turns_per_second_t setpoint) {
  return frc2::cmd::Parallel(
             // Run the shooter flywheel at the desired setpoint using
             // feedforward and feedback
             Run([this, setpoint] {
               m_shooterMotor.SetVoltage(
                   m_shooterFeedforward.Calculate(setpoint) +
                   units::volt_t(m_shooterFeedback.Calculate(
                       GetShooterVelocity().value(), setpoint.value())));
             }),
             // Wait until the shooter has reached the setpoint, and then
             // run the feeder
             frc2::cmd::WaitUntil([this] {
               return m_shooterFeedback.AtSetpoint();
             }).AndThen([this] {
               m_feederMotor.Set(ShooterConstants::kFeederSpeed);
             }))
      .WithName("Shoot");
}
